main.coffee: Fix old version warning

ed4140 opened the pull request:
Fix old version warning even when the latest is installed.
Versions were compared as strings, so '0.10.0' was less than '0.9.1'
Replaced comparison with a proper function.

Fixes https://github.com/coala/coala-atom/issues/48

ed4140 committed:
main.coffee: Fix old version warning

Fix old version warning even when the latest is installed.
Versions were compared as strings, so '0.10.0' was less than '0.9.1'
Replaced comparison with a proper function.
Closes https://github.com/coala/coala-atom/issues/48

EverWinter23 commented:
Hey, @ed4140, you are not following commit message guidelines.
Commit body lines should not exceed 72 characters.
circle-ci test is failing for this reason.

EverWinter23 commented:
Run coala with SpaceConsistencyBear before submitting.

EverWinter23 commented:
unack 0106980ac7dfe31edeec8d5a2064abf4a99996ca

ed4140 commented:
What is the problem if README.md was not changed?

RohanVB commented on lib/main.coffee:
```diff
@@ -29,7 +39,7 @@ module.exports =
     # Check version of coala
     version = helpers.exec(@executable, ['--version']).then (result) ->
-      if result <= COALA_MIN_VERSION
+      if cmpVersions(result, COALA_MIN_VERSION) < 0
```
Can we use inbuilt functions by importing from a library? Better to not reinvent the wheel by writing your own function for comparing 2 floats.  :)

ed4140 replied:
@RohanVB `'0.9.1'` is not float, comparing such objects as strings also did not work so we came up with a proper function:)

RohanVB replied:
How about ignoring the `0.` and comparing the rest of the digits which form a float? `9.1` and `10.0`, `10.1`, etc,.? All our versions at the moment follow that norm, in the future when we release 1.x we could update this to keep things simple.

ed4140 replied:
With 0.x.10 vs 0.x.9 this won't work as well. Long term solution is better

RohanVB replied:
I did a little digging around, https://www.python.org/dev/peps/pep-0440/ the current recommended method of pre/post release tagging can be handled by `pkg_resources.parse_version()`.

ed4140 replied:
@RohanVB Well done, If I was working in python I would have used this function

RohanVB replied:
Well if there's something similar you could find for js/coffee, thatd be great. Otherwise I guess we'll just use this for now.

sils replied:
so the python way of doing this would be `tuple(int(num) for num in '1.2.3'.split('.'))` (which could e put into a simple function) and then use the tuple `>` operation which would compare first elements first, then second etc.; it behaves very very well for versions:

```
>>> (1,) > (1,2)
False
>>> (1,2,3) > (1,2)
True
```

is such a thing possible for coffeescript?

RohanVB reviewed (commented):
Could you rebase and check if the `patch-2` branch is up-to-date with the current `master` branch of coala-atom, then repush? 
Also, the gitmate stuff you can get rid of by running coala locally on your project, fixing stuff there before pushing. :)

RohanVB commented:
You still have your old commits. You should squash/remove them if they aren't necessary.

ed4140 commented:
Well, I hope that now it is ok

hollamad commented:
@ed4140 Please do run `py.test` and `coala` on your local machine inside the directory from next time.
So that these gitmate reviews will be fixed before you push :)

RohanVB commented:
@ed4140 you didn't address the reviews in your code. 
Also, you need an issue reference in the commit.

@Madhukar-SH we don't have py.test tests set up on this repo yet.

RohanVB commented on lib/main.coffee:
```diff
@@ -2,6 +2,17 @@
 helpers = require 'atom-linter'
 path = require 'path'
 
+cmpVersions = (v1, v2) ->
```
Run coala locally and apply the patch on this section to fix the gitmate-bot error

RohanVB replied:
Run coala locally and apply the patch on this section to fix the gitmate-bot error

sils commented on lib/main.coffee:
```diff
@@ -2,6 +2,16 @@
 helpers = require 'atom-linter'
 path = require 'path'
 
+cmpVersions = (v1, v2) ->
```
can we have some doc comment or so? Also maybe it should be rather a function that returns a boolean, i.e. `isGreaterVersion` or so?

sils reviewed (changes requested):
also needs a rebase with conflict resolution

sils commented:
btw. big :+1: for doing this :)

ed4140 commented:
@sils Now rebased, function returns a boolean and added a comment

As for tuples - its really nice trick but unfortunately I couldn't find something like this for coffeescript

sils commented on lib/main.coffee:
```diff
@@ -2,6 +2,17 @@
 helpers = require 'atom-linter'
 path = require 'path'
 
+# This function returns true if version v2 is greater than v1
```
actually with the new function IMO the comment isn't needed anymore, this just tells the same as the name does; before I asked for it because the function name wasn't so obvious. That's a good sign for the design of the functionality :+1:

sils replied:
sorry for the back and forth though

sils commented on lib/main.coffee:
```diff
@@ -2,6 +2,17 @@
+  if minLength > 0
+    for idx in [0..minLength - 1]
+      diff = Number(v1Parts[idx]) - Number(v2Parts[idx])
+      return diff < 0 unless diff is 0
```
IMO this'd be better to understand:

```
unless v1Parts[idx] == v2Parts[idx]
  return v1Parts[idx] > v2Parts[idx]
```

esp because the `>` expression is exactly what the name says and not twirked around but with the same mathematical meaning. This leads to even easier understanding for non coffeescripters like me ;) (I hope the code sample above is even correct :P)

sils commented on lib/main.coffee:
```diff
@@ -2,6 +2,17 @@
+    for idx in [0..minLength - 1]
+      diff = Number(v1Parts[idx]) - Number(v2Parts[idx])
+      return diff < 0 unless diff is 0
+  return (v1Parts.length - v2Parts.length) < 0
```
again why don't you do `v1Parts.length < v2Parts.length`? Way easier to understand.

sils reviewed (changes requested):
Good stuff, only minor improvements needed for better understandibility

ed4140 commented:
@sils No problem, done:)

jayvdb commented on lib/main.coffee:
```diff
@@ -2,6 +2,16 @@
+    for idx in [0..minLength - 1]
+      unless Number(v1Parts[idx]) == Number(v2Parts[idx])
+        return Number(v1Parts[idx]) < Number(v2Parts[idx])
+  return v1Parts.length < v2Parts.length
```
this should be `<=`

ed4140 replied:
I did some tests and it seems this should be `=`

Tests: https://pastebin.com/rtaharqJ

ed4140 replied:
Please tell me if I'm wrong

jayvdb replied:
The problem to eliminate is `0.10.0.dev.99999` is a lower version than `0.10.0`.

`=` and `<=` are both ok.

We always use semver for real versions, so `v1Parts.length = 3` would also work.

Ohhh , but ... `0.11.0.dev.123` needs to be higher than `0.10.0`.

Doesnt coffee have a decent library for semver?? ;-)

jayvdb commented on lib/main.coffee:
```diff
@@ -2,6 +2,16 @@
 helpers = require 'atom-linter'
 path = require 'path'
 
+isGreaterVersion = (v1, v2) ->
```
this should be called `isLessThanVersion` , as 

- equal to `0.10.0` is acceptable, and
- the name of the function should be focused on the first argument (`result`)

blazeu commented:
Hello @ed4140 are you still on this?
coala-atom is not usable if this is not fixed.

You can use https://www.npmjs.com/package/semver to check for versioning instead of writing your own function.

Makman2 commented:
Closing due to inactivity.
